hs/src/Reach/Optimize.hs

```diff
@@ -520,6 +524,49 @@ instance Extract (Maybe DLVar) where
 allTheSame :: (Eq a, Sanitize a) => [a] -> Either (Maybe (a, a)) a
 allTheSame = allEqual . map sani
 
+-- Given that the result of `a` is `DLArg`, learn new bindings
+class Learn a where
+  learn :: DLArg -> a -> [(DLVar, DLArg)]
+
+instance Learn DLArg where
+  learn result = \case
+    DLA_Var dv -> [(dv, result)]
+    _ -> []
+
+instance Learn DLExpr where
+  learn result = \case
+    DLE_PrimOp _ IF_THEN_ELSE [DLA_Var dv, lhs, rhs]
+      -- r = true, x ? true : true  => []
+      -- r = v2  , x ? v2 : y       => [(x, true)]
+      | lhs == result && rhs /= result -> [(dv, true)]
+      | rhs == result && lhs /= result -> [(dv, false)]
```

Comment:
`==` may be too specific and you should use `equiv` instead (because of source location / metadata)